### PR TITLE
En_Dodongo

### PR DESCRIPTION
**Pull request by petrie911**
What other actor can give you a whole heap of bombchus?

### CODE REVIEW

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 1]
```diff
@@ -1,15 +1,52 @@
...
+extern AnimationHeader D_06004C20;
+extern AnimationHeader D_06008B1C;
+extern AnimationHeader D_060028F0;
+extern AnimationHeader D_06003088;
+extern AnimationHeader D_060013C4;
+extern AnimationHeader D_06001A44;
+
+typedef enum EnDodongoActionState {
+    DODONGO_SWEEP_TAIL,
+    DODONGO_SWALLOW_BOMB,
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:40]: i usually keep enums at the top under defines

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 2]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+    SkelAnime_ChangeAnimTransitionRepeat(&this->skelAnime, &D_06004C20, -4.0f);
+    this->actor.speedXZ = 0.0f;
+    this->timer = Math_Rand_S16Offset(30, 50);
+    this->actionState = DODONGO_IDLE;
+    EnDodongo_SetupAction(this, EnDodongo_Idle);
+}
+
+void EnDodongo_SetupWalk(EnDodongo* this) {
+    f32 frames = SkelAnime_GetFrameCount(&D_06008B1C.genericHeader);
+
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:260]: generic header can go

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 3]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+}
+
+void EnDodongo_SetupStunned(EnDodongo* this) {
+    SkelAnime_ChangeAnim(&this->skelAnime, &D_060028F0, 0.0f, 25.0f, 0.0f, 2, -4.0f);
+    this->actionState = DODONGO_STUNNED;
+    this->actor.speedXZ = 0.0f;
+    if (this->damageEffect == 0xF) {
+        this->iceTimer = 36;
+    }
+    Audio_PlayActorSound2(&this->actor, 0x389E);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:298]: sound

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 4]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+void EnDodongo_BreatheFire(EnDodongo* this, GlobalContext* globalCtx) {
+    s32 pad;
+    Vec3f velocity = { 0.0f, 0.0f, 0.0f };
+    Vec3f accel = { 0.0f, 0.0f, 0.0f };
+    Vec3f pos;
+    s16 pad2;
+    s16 fireFrame;
+
+    if ((s32)this->skelAnime.animCurrentFrame == 24) {
+        Audio_PlayActorSound2(&this->actor, 0x3801);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:325]: sound

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 5]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+    }
+    if ((29.0f <= this->skelAnime.animCurrentFrame) && (this->skelAnime.animCurrentFrame <= 43.0f)) {
+        Audio_PlayActorSound2(&this->actor, 0x3002);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:328]: sound

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 6]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+        fireFrame = this->skelAnime.animCurrentFrame - 29.0f;
+        pos = this->actor.posRot.pos;
+        pos.y += 35.0f;
+        EnDodongo_ShiftVecRadial(this->actor.posRot.rot.y, 30.0f, &pos);
+        EnDodongo_ShiftVecRadial(this->actor.posRot.rot.y, 2.5f, &accel);
+        EffectSsDFire_SpawnFixedScale(globalCtx, &pos, &velocity, &accel, 255 - (fireFrame * 10), fireFrame + 3);
+    } else if ((2.0f <= this->skelAnime.animCurrentFrame) && (this->skelAnime.animCurrentFrame <= 20.0f)) {
+        Audio_PlayActorSound2(&this->actor, 0x3058);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:336]: not gonna comment on any more sounds

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 7]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+                    func_8002836C(globalCtx, &pos, &deathFireVel, &deathFireAccel, &this->bombSmokePrimColor,
+                                  &this->bombSmokeEnvColor, 400, 10, 10);
+                }
+                Audio_PlayActorSound2(&this->actor, 0x180E);
+                func_8003426C(&this->actor, 0x4000, 0x78, 0, 8);
+            }
+        }
+    }
+    if ((s32)this->skelAnime.animCurrentFrame < 28) {
+        if (((s32)this->skelAnime.animCurrentFrame < 26) && (this->timer < 11)) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:397]: <= 10

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 8]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+    }
+}
+
+void EnDodongo_Walk(EnDodongo* this, GlobalContext* globalCtx) {
+    s32 pad;
+    f32 playbackSpeed;
+    Player* player = PLAYER;
+    s16 yawDiff = (s16)(this->actor.yawTowardsLink - this->actor.shape.rot.y);
+
+    yawDiff = (yawDiff < 0) ? -yawDiff : yawDiff;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:425]: abs should work

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 9]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+
+    if (Math_Vec3f_DistXZ(&this->actor.initPosRot.pos, &player->actor.posRot.pos) < 400.0f) {
+        Math_SmoothScaleMaxMinS(&this->actor.posRot.rot.y, this->actor.yawTowardsLink, 1, 0x1F4, 0);
+        this->actor.flags |= 1;
+        if ((this->actor.xzDistFromLink < 100.0f) && (yawDiff < 0x1388) && (this->actor.yDistFromLink < 60.0f)) {
+            EnDodongo_SetupBreatheFire(this);
+        }
+    } else {
+        this->actor.flags &= ~1;
+        if ((150.0f < Math_Vec3f_DistXZ(&this->actor.posRot.pos, &this->actor.initPosRot.pos)) ||
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:464]: flip 150 to other side if possible

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 10]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+    if (SkelAnime_FrameUpdateMatrix(&this->skelAnime) != 0) {
+        if (this->timer == 0) {
+            bomb = (EnBom*)Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_EN_BOM, this->actor.posRot.pos.x,
+                                       this->actor.posRot.pos.y, this->actor.posRot.pos.z, 0, 0, 6, BOMB_BODY);
+            if (bomb != NULL) {
+                bomb->timer = 0;
+                this->timer = 8;
+            }
+        }
+    } else if ((s32)this->skelAnime.animCurrentFrame == 0x34) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:583]: frame conunt decimal

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 11]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+        Audio_PlayActorSound2(&this->actor, 0x387A);
+    }
+    if (this->timer != 0) {
+        this->timer--;
+        if (this->timer == 0) {
+            Item_DropCollectibleRandom(globalCtx, &this->actor, &this->actor.posRot.pos, 0x40);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:589]: i think there might be an enum for the last arg

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 12]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+        CollisionCheck_SetAT(globalCtx, &globalCtx->colChkCtx, &this->colliderAT.base);
+    }
+    this->actor.posRot2.pos.x = this->actor.posRot.pos.x + Math_Sins(this->actor.shape.rot.y) * -30.0f;
+    this->actor.posRot2.pos.y = this->actor.posRot.pos.y + 20.0f;
+    this->actor.posRot2.pos.z = this->actor.posRot.pos.z + Math_Coss(this->actor.shape.rot.y) * -30.0f;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dodongo/func_809F9A80.s")
+s32 EnDodongo_OverrideLimbDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3f* pos, Vec3s* rot,
+                               Actor* thisx) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:691]: void* thisx

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 13]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+    EnDodongo* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dodongo/func_809F9AF8.s")
+    if ((limbIndex == 0xF) || (limbIndex == 0x10)) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:694]: decimal limbs

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 14]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+        Matrix_Scale(this->bodyScale.x, this->bodyScale.y, this->bodyScale.z, MTXMODE_APPLY);
+    }
+    return 0;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dodongo/func_809F9C3C.s")
+void EnDodongo_PostLimbDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3s* rot, Actor* thisx) {
+    static Vec3f legOffsets[3] = {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:701]: void* thisx

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 15]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+    Vec3f headOffset = { 1500.0f, 300.0f, 0.0f };
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dodongo/func_809F9CA0.s")
+    func_800628A4(limbIndex, &this->colliderBody);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dodongo/func_809F9DC8.s")
+    switch (limbIndex) {
+        case 2:
+            if ((this->actionState == DODONGO_BREATHE_FIRE) && (29.0f < this->skelAnime.animCurrentFrame) &&
+                (this->skelAnime.animCurrentFrame < 43.0f)) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:738]: id flip 29 if possible

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 16]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
+                Matrix_MultVec3f(&tailTipOffset, &tailTip);
+                Matrix_MultVec3f(&baseOffset, &tailBase);
+                EffectBlure_AddVertex(Effect_GetByIndex(this->blureIdx), &tailTip, &tailBase);
+            } else if ((this->actionState == DODONGO_SWEEP_TAIL) && (this->timer != 0)) {
+                EffectBlure_AddSpace(Effect_GetByIndex(this->blureIdx));
+            }
+            break;
+        case 21:
+            Matrix_MultVec3f(&legOffsets[1], &this->leftFootPos);
+            break; // legOffsets[1] is a zero vector
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:765]: comment not needed imo

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 17]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dodongo/EnDodongo_Update.s")
+void EnDodongo_Draw(Actor* thisx, GlobalContext* globalCtx2) {
+    GlobalContext* globalCtx = globalCtx2;
+    EnDodongo* this = THIS;
+    s32 index;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dodongo/func_809FA0F8.s")
+    func_80093D18(globalCtx->state.gfxCtx);
+    SkelAnime_DrawOpa(globalCtx, this->skelAnime.skeleton, this->skelAnime.limbDrawTbl, EnDodongo_OverrideLimbDraw,
+                   EnDodongo_PostLimbDraw, &this->actor);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:815]: can pass this instead

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 18]
```diff
@@ -21,59 +58,803 @@ const ActorInit En_Dodongo_InitVars = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dodongo/func_809FA14C.s")
+    if (this->iceTimer != 0) {
+        this->actor.dmgEffectTimer++;
+        if (1) {}
+        this->iceTimer--;
+        if ((this->iceTimer % 4) == 0) {
+            index = this->iceTimer >> 2;
+            EffectSsEnIce_SpawnFlyingVec3f(globalCtx, &this->actor, &this->icePos[index], 0x96, 0x96, 0x96, 0xFA, 0xEB,
+                                           0xF5, 0xFF, 1.8f);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:824]: decimal args when applicable

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.h [hunk 19]
```diff
@@ -6,11 +6,45 @@
...
+    /* 0x031C */ Vec3f leftFootPos;
+    /* 0x0328 */ Vec3f rightFootPos;
+    /* 0x0334 */ Vec3f mouthPos;
+    /* 0x0334 */ Vec3f headPos;
+    /* 0x034C */ Vec3f bodyScale;
+    /* 0x0358 */ Vec3f icePos[9];
+    /* 0x03C4 */ Color_RGBA8 bombSmokePrimColor;
+    /* 0x03C8 */ Color_RGBA8 bombSmokeEnvColor; 
+    /* 0x03CC */ u8 damageEffect;
+    /* 0x03CD */ char unk_3CD[3];
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.h:34]: 3 bytes of padding can go

**Review by fig02 — COMMENTED** [review]
i didnt comment on all of them but there are alot of sfx ids that need to be added

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 20]
```diff
@@ -1,15 +1,52 @@
...
+#include "overlays/actors/ovl_En_Bom/z_en_bom.h"
+#include "overlays/actors/ovl_En_Bombf/z_en_bombf.h"
 
 #define FLAGS 0x00000015
 
 #define THIS ((EnDodongo*)thisx)
 
 void EnDodongo_Init(Actor* thisx, GlobalContext* globalCtx);
-void EnDodongo_Destroy(Actor* thisx, GlobalContext* globalCtx);
+void EnDodongo_Destroy(Actor* thix, GlobalContext* globalCtx);
```
[the remark this answers is not in the document]
Comment (petrie911) [reply src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:10]: thicc

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c (hunk 11 above)
Comment (petrie911) [reply src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:589]: Not for the random ones as far as I know.

Comment (fig02) [reply src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:589]: ah true

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c (hunk 15 above)
Comment (petrie911) [reply src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:736]: Went with something a little different. As you know, I do prefer the ` a < x && x < b` ordering.

**Review by fig02 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 21]
```diff
@@ -21,59 +58,801 @@ const ActorInit En_Dodongo_InitVars = {
...
+            blureInit.p1EndColor[1] = blureInit.p1EndColor[2] = blureInit.p2EndColor[0] = blureInit.p2EndColor[1] =
+                blureInit.p2EndColor[2] = 255;
+
+    blureInit.p1EndColor[3] = blureInit.p2EndColor[3] = 0;
+    blureInit.p2StartColor[3] = 64;
+    blureInit.elemDuration = 8;
+    blureInit.unkFlag = false;
+    blureInit.calcMode = 2;
+
+    Effect_Add(globalCtx, &this->blureIdx, 1, 0, 0, &blureInit);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:232]:
```suggestion
    Effect_Add(globalCtx, &this->blureIdx, EFFECT_BLURE1, 0, 0, &blureInit);
```

File: src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c [hunk 22]
```diff
@@ -21,59 +58,801 @@ const ActorInit En_Dodongo_InitVars = {
...
+            s16 yawToHome = Math_Vec3f_Yaw(&this->actor.posRot.pos, &this->actor.initPosRot.pos);
+
+            Math_SmoothScaleMaxMinS(&this->actor.posRot.rot.y, yawToHome, 1, 0x1F4, 0);
+        }
+        if (this->retreatTimer != 0) {
+            this->retreatTimer--;
+        }
+        this->timer--;
+        if (this->timer == 0) {
+            if (0.7f < Math_Rand_ZeroOne()) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dodongo/z_en_dodongo.c:472]:
```suggestion
            if (Math_Rand_ZeroOne() > 0.7f) {
```

**Review by Roman971 — APPROVED** [review]

**Pull request merged**